src/Planner/CollectColumnIdentifiers.cpp

```diff
@@ -35,6 +39,30 @@ class CollectTopLevelColumnIdentifiersVisitor : public ConstInDepthQueryTreeVisi
 
     void visitImpl(const QueryTreeNodePtr & node)
     {
+        if (node->getNodeType() == QueryTreeNodeType::FUNCTION)
```

Comment:
It's a bit strange that we traverse scalar correlated subqueries only as a function argument.
E.g. will it work if I write `select ... where (correlated sclar subquery)`?

Maybe it's just better to allow query/union in `needChildVisit`.